maro/rl_v3/policy_trainer/distributed_discrete_maddpg.py

```diff
@@ -0,0 +1,532 @@
+
+    def _record_impl(self, transition_batch: MultiTransitionBatch) -> None:
+        self._replay_memory.put(transition_batch)
+
+    def _register_policies_impl(self, policies: List[RLPolicy]) -> None:
+        assert all(isinstance(policy, DiscretePolicyGradient) for policy in policies)
+
+        self._num_policies = len(policies)
+        self._policy_names = [policy.name for policy in policies]
+
+        if self._shared_critic:
+            self._q_critic_net = self._get_q_critic_net_func()
+            self._q_critic_net.to(self._device)
+            self._target_q_critic_net = clone(self._q_critic_net)
+            self._target_q_critic_net.to(self._device)
+            self._target_q_critic_net.eval()
+        else:
+            self._q_critic_net = None
+            self._target_q_critic_net = None
+
+        self._workers: List[DiscreteMADDPGWorker] = []
+        self._worker_indexes: List[List[int]] = []
+        cursor = 0
+        while cursor < self.num_policies:
+            worker = DiscreteMADDPGWorker(
+                reward_discount=self._reward_discount, get_q_critic_net_func=self._get_q_critic_net_func,
+                shared_critic=self._shared_critic, device=self._device_str, data_parallel=self._data_parallel,
+                critic_loss_coef=self._critic_loss_coef, soft_update_coef=self._soft_update_coef,
+                update_target_every=self._update_target_every, q_value_loss_func=self._q_value_loss_func
+            )
+
+            cursor_end = min(cursor + self._group_size, self.num_policies)
+            indexes = list(range(cursor, cursor_end))
+            cursor = cursor_end
+            worker.register_policies({i: policies[i] for i in indexes})
+
+            self._workers.append(worker)
+            self._worker_indexes.append(indexes)
+
+        # Replay
+        self._replay_memory = RandomMultiReplayMemory(
+            capacity=self._replay_memory_capacity,
+            state_dim=self._get_q_critic_net_func().state_dim,
+            action_dims=[policy.action_dim for policy in policies],
+            agent_states_dims=[policy.state_dim for policy in policies]
+        )
+
+    def _get_batch(self, batch_size: int = None) -> MultiTransitionBatch:
+        return self._replay_memory.sample(batch_size if batch_size is not None else self._train_batch_size)
+
+    def _train_step_impl(self) -> None:
+        for _ in range(self._num_epoch):
+            self._improve(self._get_batch())
+
+    def _improve(self, batch: MultiTransitionBatch) -> None:
+        for worker in self._workers:
+            worker.set_batch(batch)
+
+        # Collect next actions
+        next_action_dict: Dict[int, torch.Tensor] = {}
+        for worker in self._workers:
+            next_action_dict.update(worker.get_target_action_dict())
+        next_actions = [next_action_dict[i] for i in range(self.num_policies)]
+
+        # Update critic
+        if self._shared_critic:
+            grads = self._get_batch_grad(
```

Comment:
I suggest that leave all the critic & actor update methods in the `MADDPGWorker` to decouple.
If `_shared_critic`, the trainer could maintain a worker locally to update shared critic.